numerics/polynomial_evaluators_body.hpp

```diff
@@ -0,0 +1,221 @@
+#pragma once
+
+#include "numerics/polynomial_evaluators.hpp"
+
+#include <tuple>
+
+namespace principia {
+namespace numerics {
+namespace internal_polynomial_evaluators {
+
+namespace {
+
+// Greatest power of 2 less than or equal to n.  8 -> 8, 7 -> 4.
+constexpr int flp2(int const n) {
+  return n == 0 ? 0 : n == 1 ? 1 : flp2(n >> 1) << 1;
+}
+
+// Ceiling log2 of n.  8 -> 3, 7 -> 2.
+constexpr int log2(int const n) {
+  return n == 1 ? 0 : log2(n >> 1) + 1;
+}
+
+}  // namespace
+
+// Generator for repeated squaring:
+//   SquareGenerator<Length, 0>::Type is Length
+//   SquareGenerator<Length, 1>::Type is Exponentiation<Length, 2>
+//   SquareGenerator<Length, 2>::Type is Exponentiation<Length, 4>
+//   SquareGenerator<Length, n>::Type is Exponentiation<Length, 2^n>
+// etc.
+template<typename Argument, int n>
+struct SquareGenerator {
+  using Type = Square<typename SquareGenerator<Argument, n - 1>::Type>;
+  static Type Evaluate(Argument const& argument);
+};
+template<typename Argument>
+struct SquareGenerator<Argument, 0> {
+  using Type = Argument;
+  static Type Evaluate(Argument const& argument);
+};
+
+template<typename Argument, typename>
+struct SquaresGenerator;
+template<typename Argument, int... orders>
+struct SquaresGenerator<Argument, std::integer_sequence<int, orders...>> {
+  using Type = std::tuple<typename SquareGenerator<Argument, orders>::Type...>;
+  static Type Evaluate(Argument const& argument);
+};
+
+template<typename Argument, int n>
+auto SquareGenerator<Argument, n>::Evaluate(Argument const& argument) -> Type {
+  auto const argument_n_minus_1 =
+      SquareGenerator<Argument, n - 1>::Evaluate(argument);
+  return argument_n_minus_1 * argument_n_minus_1;
+}
+
+template<typename Argument>
+auto SquareGenerator<Argument, 0>::Evaluate(Argument const& argument) -> Type {
+  return argument;
+}
+
+template<typename Argument, int... orders>
+auto SquaresGenerator<Argument, std::integer_sequence<int, orders...>>::
+    Evaluate(Argument const& argument) -> Type {
+  return std::make_tuple(
+      SquareGenerator<Argument, orders>::Evaluate(argument)...);
+}
+
+template<typename Value, typename Argument, int degree, int low, int high>
+struct InternalEstrinEvaluator {
```

Comment:
This needs comment: what does the `Evaluate` function evaluate? what are `low` and `high`?

Reply:
Done.